syntax = "proto3";
package reflex;

service ReflexService {
  // --- Control Plane ---
  // 1. Get Physics State (Internal Mechanics)
  rpc GetPhysics (Empty) returns (PhysicsResponse);

  // 2. Get OODA State (Decision Logic)
  rpc GetOODA (Empty) returns (OODAResponse);

  // 3. Trigger Veto (Nuclear Option)
  rpc TriggerVeto (VetoRequest) returns (Ack);

  // 4. Demote Provisional (Safety Staircase)
  rpc DemoteProvisional (DemoteRequest) returns (Ack);

  // --- Legacy / Aux ---
  // Force a risk level change
  rpc TriggerRatchet (RatchetRequest) returns (Ack);
  // Update system config
  rpc UpdateConfig (ConfigPayload) returns (Ack);
  // Stream heartbeat (Legacy, deprecated by WS)
  rpc GetStream (Empty) returns (stream Heartbeat);
}

// --- Messages ---

message PhysicsResponse {
    double price = 1;
    double velocity = 2;
    double acceleration = 3;
    double jerk = 4;
    double entropy = 5;
    double efficiency_index = 6;
    double timestamp = 7;
}

message OODAResponse {
    PhysicsResponse physics = 1;
    optional double sentiment_score = 2;
    optional string nearest_regime = 3;
    string decision = 4; // BUY, SELL, HOLD
    map<string, double> weights = 5; // Simons, Kepler, Hypatia weights
}

message VetoRequest {
    string reason = 1;
    string operator = 2;
}

message DemoteRequest {
    string reason = 1;
    string target_level = 2; // e.g., "Q1", "Floor"
}

message RatchetRequest {
  enum Level {
    IDLE = 0;
    TIGHTEN = 1; // Reduce size
    FREEZE = 2;  // Cancel all, hold cash
    KILL = 3;    // Liquidate everything, shutdown
  }
  Level level = 1;
  string reason = 2;
}

message ConfigPayload {
  string key = 1;
  double value = 2;
}

message Heartbeat {
  double timestamp = 1;
  bool brain_connected = 2;
  double efficiency_index = 3;
  double p_riemann = 4;
}

message Ack {
  bool success = 1;
  string message = 2;
}

message Empty {}